File: src/brfc/Time.cpp
/*
Copyright 2010-2011 Estonian Meteorological and Hydrological Institute

This file is part of baltrad-db.

baltrad-db is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

baltrad-db is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with baltrad-db. If not, see <http://www.gnu.org/licenses/>.
*/

#include <brfc/Time.hpp>

#include <brfc/exceptions.hpp>
#include <brfc/DateTime.hpp>
#include <brfc/TimeDelta.hpp>
#include <brfc/util/iso8601.hpp>

namespace brfc {

Time::Time(int hour,
           int minute,
           int second,
           int msec)
        : msec_(0) {
    msec_ = hour * MSECS_IN_HOUR +
            minute * MSECS_IN_MIN +
            second * MSECS_IN_SEC +
            msec;
    if (hour < 0 or hour >= HOURS_IN_DAY)
        throw std::invalid_argument("hour value out of range");
    if (minute < 0 or minute >= MINS_IN_HOUR)
        throw std::invalid_argument("minute value out of range");
    if (second < 0 or second >= SECS_IN_MIN)
        throw std::invalid_argument("second value out of range");
    if (msec < 0 or msec >= MSECS_IN_SEC)
        throw std::invalid_argument("msec value out of range");
    if (msec_ >= MSECS_IN_DAY)
        throw std::invalid_argument("time value out of range");
}

Time::Time(unsigned int msec)
        : msec_(msec) {
    if (msec_ >= MSECS_IN_DAY)
        throw std::invalid_argument("time value out of range");
}

Time::Time(const Time& other)
        : msec_(other.msec_) {
}

Time&
Time::operator=(const Time& rhs) {
    if (this != &rhs) {
        msec_ = rhs.msec_;
    }
    return *this;
}

Time
Time::now() {
    return DateTime::now().time();
}

Time
Time::utc_now() {
    return DateTime::utc_now().time();
}

int
Time::hour() const {
    return msec_ / MSECS_IN_HOUR;
}

int
Time::minute() const {
    return (msec_ % MSECS_IN_HOUR) / MSECS_IN_MIN;
}

int
Time::second() const {
    return (msec_ / MSECS_IN_SEC) % SECS_IN_MIN;
}

int
Time::msec() const {
    return msec_ % MSECS_IN_SEC;
}

Time&
Time::hour(int _hour) {
    if (_hour < 0 or _hour >= HOURS_IN_DAY)
        throw std::invalid_argument("invalid hour");
    msec_ = msec_ - (hour() * MSECS_IN_HOUR) + (_hour * MSECS_IN_HOUR);
    return *this;
}

Time&
Time::minute(int _minute) {
    if (_minute < 0 or _minute >= MINS_IN_HOUR)
        throw std::invalid_argument("invalid minute");
    msec_ = msec_ - (minute() * MSECS_IN_MIN) + (_minute * MSECS_IN_MIN);
    return *this;
}

Time&
Time::second(int _second) {
    if (_second < 0 or _second >= SECS_IN_MIN)
        throw std::invalid_argument("invalid second");
    msec_ = msec_ - (second() * MSECS_IN_SEC) + (_second * MSECS_IN_SEC);
    return *this;
}

Time&
Time::msec(int _msec) {
    if (_msec < 0 or _msec >= MSECS_IN_SEC)
        throw std::invalid_argument("invalid msec");
    msec_ = msec_ - msec() + _msec;
    return *this;
}

Time&
Time::operator+=(const TimeDelta& td) {
    // td.msecs() is always positive
    msec_ = (td.msecs() + msec_) % MSECS_IN_DAY;
    return *this;
}

Time
Time::operator+(const TimeDelta& td) const {
    return Time(*this) += td;
}

Time
Time::from_iso_string(const std::string& str) {
    iso8601::time t;
    if (not iso8601::parse_time(str, t))
        throw std::invalid_argument("invalid ISO8601 time: " + str);
    return Time(t.hour, t.minute, t.second, t.usec / 1000);
}

std::string
Time::to_iso_string(bool extended) const {
    iso8601::time t(hour(), minute(), second(), msec() * 1000);
    if (extended)
        return iso8601::to_string(t, iso8601::format::EXTENDED);
    else
        return iso8601::to_string(t, iso8601::format::BASIC);
}

} // namespace brfc

File: client/java/src/eu/baltrad/bdb/util/Time.java
/*
Copyright 2010-2011 Estonian Meteorological and Hydrological Institute

This file is part of baltrad-db.

baltrad-db is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

baltrad-db is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with baltrad-db. If not, see <http://www.gnu.org/licenses/>.
*/

package eu.baltrad.bdb.util;

import org.joda.time.DateTimeZone;
import org.joda.time.LocalTime;
import org.joda.time.format.DateTimeFormatter;
import org.joda.time.format.DateTimeFormatterBuilder;

public final class Time {
  private static DateTimeFormatter isoFormatter;
  private static DateTimeFormatter extendedIsoFormatter;

  static {
    DateTimeFormatterBuilder builder = new DateTimeFormatterBuilder();
    isoFormatter = builder.appendHourOfDay(2)
                          .appendMinuteOfHour(2)
                          .appendSecondOfMinute(2)
                          .toFormatter();
    builder.clear();
    extendedIsoFormatter = builder.appendHourOfDay(2)
                                  .appendLiteral(':')
                                  .appendMinuteOfHour(2)
                                  .appendLiteral(':')
                                  .appendSecondOfMinute(2)
                                  .toFormatter();
  }

  private final LocalTime value;
  
  /**
   * Construct using current UTC time.
   */
  public Time() {
    value = new LocalTime(DateTimeZone.UTC);
  }

  public Time(int hour, int minute, int second) {
    this(hour, minute, second, 0);
  }

  public Time(int hour, int minute, int second, int msec) {
    value = new LocalTime(hour, minute, second, msec);
  }

  protected Time(LocalTime value) {
    this.value = value;
  }

  public static Time fromIsoString(String value) {
    return new Time(isoFormatter.parseLocalTime(value));
  }

  public int hour() { return value.getHourOfDay(); }
  public int minute() { return value.getMinuteOfHour(); }
  public int second() { return value.getSecondOfMinute(); }
  public int msec() { return value.getMillisOfSecond(); }
  public int getCumulativeMsecs() { return value.getMillisOfDay(); }
  
  public String toIsoString() {
    return isoFormatter.print(value);
  }

  public String toExtendedIsoString() {
    return extendedIsoFormatter.print(value);
  }

  @Override
  public String toString() {
    return toIsoString();
  }

  @Override
  public boolean equals(Object other) {
    if (this == other)
      return true;
    
    if (!(other instanceof Time))
      return false;
    
    Time that = (Time)(other);
    return this.value.equals(that.value);
  }

  @Override
  public int hashCode() {
    return this.value.hashCode();
  }
}
